src/modules/mavlink/mavlink_stream.cpp

```diff
@@ -58,73 +58,53 @@ MavlinkStream::update(const hrt_abstime &t)
 {
 	update_data();
 
-	// If the message has never been sent before we want
-	// to send it immediately and can return right away
-	if (_last_sent == 0) {
-		// this will give different messages on the same run a different
-		// initial timestamp which will help spacing them out
-		// on the link scheduling
+	bool sent = false;
+
+	if (_interval == 0) {
+		// We don't need to send anything if the inverval is 0. send() will be called manually.
+		return 0;
+
+	} else if (_last_sent == 0) {
+		// If the message has never been sent before we want
+		// to send it immediately and can return right away
 		if (send()) {
 			_last_sent = hrt_absolute_time();
+			sent = true;
+		}
+
+	} else if (_interval < 0) {
+		// unlimited rate
+		if (send()) {
+			_last_sent = t;
+			sent = true;
+		}
 
-			if (!_first_message_sent) {
-				_first_message_sent = true;
+	} else if (const_rate()) {
+		if (t >= _last_sent + _interval) {
+			if (send()) {
+				_last_sent = math::constrain(_last_sent + _interval, t - _interval, t);
+				sent = true;
 			}
 		}
 
-		return 0;
-	}
+	} else {
+		int interval = _interval * _mavlink->get_rate_div();
 
-	// One of the previous iterations sent the update
-	// already before the deadline
-	if (_last_sent > t) {
-		return -1;
+		if (t >= _last_sent + interval) {
+			if (send()) {
+				_last_sent = math::constrain(_last_sent + interval, t - interval, t);
```

Comment:
The `t >= _last_sent + interval` check already ensures the last condition:
```suggestion
				_last_sent = math::max(_last_sent + interval, t - interval);
```

Reply:
Thanks.